packages/near-membrane-base/src/membrane.ts

```diff
@@ -0,0 +1,1549 @@
+    ): Parameters<CallableDescriptorCallback> {
+        const { configurable, enumerable, writable, value, get, set } = partialDesc;
+        return [
+            'configurable' in partialDesc ? !!configurable : undefinedSymbol,
+            'enumerable' in partialDesc ? !!enumerable : undefinedSymbol,
+            'writable' in partialDesc ? !!writable : undefinedSymbol,
+            'value' in partialDesc ? getTransferableValue(value) : undefinedSymbol,
+            'get' in partialDesc ? getTransferableValue(get) : undefinedSymbol,
+            'set' in partialDesc ? getTransferableValue(set) : undefinedSymbol,
+        ];
+    }
+
+    function lockShadowTarget(shadowTarget: ShadowTarget, targetPointer: Pointer) {
+        copyForeignDescriptorsIntoShadowTarget(shadowTarget, targetPointer);
+        const protoPointer = foreignCallableGetPrototypeOf(targetPointer);
+        // setting up __proto__ of the shadowTarget
+        setPrototypeOf(shadowTarget, getLocalValue(protoPointer));
+        // locking down the extensibility of shadowTarget
+        preventExtensions(shadowTarget);
+    }
+
+    class BoundaryProxyHandler implements ProxyHandler<ShadowTarget> {
+        // public fields
+        revoke: () => void;
+
+        proxy: ShadowTarget;
+
+        ownKeys: ProxyHandler<ShadowTarget>['ownKeys'];
+
+        isExtensible: ProxyHandler<ShadowTarget>['isExtensible'];
+
+        getOwnPropertyDescriptor: ProxyHandler<ShadowTarget>['getOwnPropertyDescriptor'];
+
+        getPrototypeOf: ProxyHandler<ShadowTarget>['getPrototypeOf'];
+
+        get: ProxyHandler<ShadowTarget>['get'];
+
+        set: ProxyHandler<ShadowTarget>['set'];
+
+        has: ProxyHandler<ShadowTarget>['has'];
+
+        setPrototypeOf: ProxyHandler<ShadowTarget>['setPrototypeOf'];
+
+        deleteProperty: ProxyHandler<ShadowTarget>['deleteProperty'];
+
+        preventExtensions: ProxyHandler<ShadowTarget>['preventExtensions'];
+
+        defineProperty: ProxyHandler<ShadowTarget>['defineProperty'];
+
+        // the purpose of this public field is to help developers to identify what side of the
+        // membrane they are debugging.
+        readonly color = color;
+
+        // callback to prepare the foreign realm before any operation
+        private readonly targetPointer: Pointer;
+
+        // apply trap is generic, and should never change independently of the type of membrane
+        readonly apply = function applyTrap(
+            this: BoundaryProxyHandler,
+            _shadowTarget: ShadowTarget,
+            thisArg: any,
+            args: any[]
+        ): any {
+            const { targetPointer } = this;
+            const thisArgValueOrPointer = getTransferableValue(thisArg);
+            const listOfValuesOrPointers = args.map(getTransferableValue);
+            const foreignValueOrCallable = foreignCallableApply(
+                targetPointer,
+                thisArgValueOrPointer,
+                ...listOfValuesOrPointers
```

Comment:
☝️ Spreading an array invoked the `Array.prototype[Symbol.iterator]` and should be avoided.